### ISSUE TITLE
Why diff results from custom objective vs built-in logitraw obj?

### ISSUE THREAD

**Issue by akhil137** [dmlc/xgboost#1373]
The custom obj. example should recreate results from the built-in objective "binary:logitraw" if I understand it.  However the code below shows a difference.  Why is that?  (it doesn't seem to have to do with the eval_metric function, as one can just pass the objective parameter in the custom model below and reach the same conclusion).

R Code below:

```
#Code below to demo difference between custom objective and builtin logistic
library(xgboost)
#generate synthetic data
set.seed(1979)
#num units/records
n=100
#num covariates
ncov=4
z=matrix(replicate(n,rnorm(ncov)),nrow=n)
alpha=c(-1,0.5,-0.25,-0.1)
za=z%*%alpha
p=exp(za)/(1+exp(za))
t=rbinom(n,1,p)
dtrain<-xgb.DMatrix(data=z,label=t)


#Custom objective and error 
logregobj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  preds <- 1/(1 + exp(-preds))
  grad <- preds - labels
  hess <- preds * (1 - preds)
  return(list(grad = grad, hess = hess))
}

evalerror <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  err <- as.numeric(sum(labels != (preds > 0)))/length(labels)
  return(list(metric = "error", value = err))
}

#Create model based on custom objective above
param<-list(objective=logregobj, eval_metric=evalerror)
bstCustom <- xgb.train(param,dtrain, nrounds=10)

#Create model based on built-in objective
bstCheck <- xgb.train(list(objective="binary:logitraw"),dtrain, nrounds=10)
#Error between two methods larger than expected (22.40854)
sum(abs(predict(bstCustom,z)-predict(bstCheck,z)))
```
# sessionInfo()

R version 3.3.0 (2016-05-03)
Platform: x86_64-apple-darwin15.4.0 (64-bit)
Running under: OS X 10.11.5 (El Capitan)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] xgboost_0.4-4

loaded via a namespace (and not attached):
[1] magrittr_1.5     Matrix_1.2-6     tools_3.3.0      stringi_1.1.1    grid_3.3.0       data.table_1.9.6
[7] stringr_1.0.0    chron_2.3-47     lattice_0.20-33

**Comment by Laurae2** [comment]
You are not using the right objective function. One is training for linear regression (apparently, as it is the default) while the other is training for logistic regression. Please see the revised code below, which leads to 0.

```
library(xgboost)
#generate synthetic data
set.seed(1979)
#num units/records
n=100
#num covariates
ncov=4
z=matrix(replicate(n,rnorm(ncov)),nrow=n)
alpha=c(-1,0.5,-0.25,-0.1)
za=z%*%alpha
p=exp(za)/(1+exp(za))
t=rbinom(n,1,p)
dtrain<-xgb.DMatrix(data=z,label=t)


#Custom objective and error 
logregobj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  preds <- 1/(1 + exp(-preds))
  grad <- preds - labels
  hess <- preds * (1 - preds)
  return(list(grad = grad, hess = hess))
}

evalerror <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  preds <- 1/(1 + exp(-preds))
  err <- as.numeric(sum(labels != (preds > 0)))/length(labels)
  return(list(metric = "error", value = err))
}

#Create model based on custom objective above
set.seed(11111)
param<-list(objective=logregobj)
bstCustom <- xgb.train(list(objective="binary:logitraw", obj=logregobj, feval=evalerror), dtrain, nrounds = 10, nthread = 1, verbose = 1, watchlist = list(train = dtrain))

#Create model based on built-in objective
set.seed(11111)
bstCheck <- xgb.train(list(objective="binary:logitraw", feval=evalerror), dtrain, nrounds = 10, nthread = 1, verbose = 1, watchlist = list(train = dtrain))

#Identical result (0)
sum(abs(predict(bstCustom,z)-predict(bstCheck,z)))

```

**Comment by akhil137** [comment]
I consider it a bug that a custom objective written for a logistic loss function matches exactly the results of objective="reg:linear" (squared loss) rather than objective="binary:logitraw".  

Thanks for the insight that a custom objective loss function is not in fact being obeyed, but rather the default loss function (squared loss).

**Comment by Laurae2** [comment]
@akhil137 It shouldn't give that result, because the documentation I had, used "obj" for the objective (not sure if it is fixed - because if inputting no "objective", it defaults to linear). However, when you use "objective" and use RMSE as the metric to observe, you get the results on the last code snip which clearly do not match what was before.

Here are more snippet results. Sigmoid (logistic) should decrease using RMSE (expected and valid), but logitraw (before Sigmoid) should increase (which is the case here). But why they increase differently is strange.

```
> bstCustom <- xgb.train(param,dtrain, nrounds=10, watchlist=list(train=dtrain))
[0]     train-rmse:0.358591
[1]     train-rmse:0.430003
[2]     train-rmse:0.579836
[3]     train-rmse:0.699472
[4]     train-rmse:0.832016
[5]     train-rmse:0.943286
[6]     train-rmse:1.067643
[7]     train-rmse:1.155419
[8]     train-rmse:1.244043
[9]     train-rmse:1.323134

> bstCheck <- xgb.train(list(objective="binary:logistic", eval_metric="rmse"),dtrain, nrounds=10, watchlist=list(train=dtrain))
[0]     train-rmse:0.444926
[1]     train-rmse:0.410217
[2]     train-rmse:0.379775
[3]     train-rmse:0.346842
[4]     train-rmse:0.325672
[5]     train-rmse:0.299801
[6]     train-rmse:0.284048
[7]     train-rmse:0.271046
[8]     train-rmse:0.258477
[9]     train-rmse:0.246660

> bstCheck <- xgb.train(list(objective="binary:logitraw", eval_metric="rmse"),dtrain, nrounds=10, watchlist=list(train=dtrain))
[0]     train-rmse:0.629228
[1]     train-rmse:0.658737
[2]     train-rmse:0.711979
[3]     train-rmse:0.773183
[4]     train-rmse:0.870353
[5]     train-rmse:0.993272
[6]     train-rmse:1.112008
[7]     train-rmse:1.215678
[8]     train-rmse:1.298016
[9]     train-rmse:1.404101
```
